### PR TITLE
[SPARK-7269] [SQL] Incorrect analysis for aggregation(use semanticEquals)

### PR DESCRIPTION
**Pull request by cloud-fan**
A modified version of https://github.com/apache/spark/pull/6110, use `semanticEquals` to make it more efficient.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @chenghao-intel @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -76,6 +76,12 @@ abstract class Expression extends TreeNode[Expression] {
       case u: UnresolvedAttribute => PrettyAttribute(u.name)
     }.toString
   }
+
+  def semanticEquals(other: Expression): Boolean = this.getClass == other.getClass &&
+    this.productIterator.zip(other.asInstanceOf[Product].productIterator).forall {
+      case (e1: Expression, e2: Expression) => e1 semanticEquals e2
+      case (i1, i2) => i1 == i2
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: What about `case class Coalesce(children: Seq[Expression])`?

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: The difficulty here is we probably never knows `semanticEquals` in a general way, that's why I said we need to re-implemented for lots of expressions if we added this.

**Comment by cloud-fan** [comment]
Hi @chenghao-intel , as far as I know, we can only instantiate leaf expressions which are all case classes. So probably we can design the `semanticEquals` as how case class implement `equals` but use `semanticEquals` for expression elements.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 2]
```diff
@@ -76,6 +76,15 @@ abstract class Expression extends TreeNode[Expression] {
       case u: UnresolvedAttribute => PrettyAttribute(u.name)
     }.toString
   }
+
+  def semanticEquals(other: Expression): Boolean = this.getClass == other.getClass && {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: scala doc please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -148,17 +147,17 @@ class Analyzer(
      * @param exprs the attributes in sequence
      * @return the attributes of non selected specified via bitmask (with the bit set to 1)
      */
-    private def buildNonSelectExprSet(bitmask: Int, exprs: Seq[Expression])
-    : OpenHashSet[Expression] = {
-      val set = new OpenHashSet[Expression](2)
+    private def buildNonSelectExprs(bitmask: Int, exprs: Seq[Expression])
+    : Seq[Expression] = {
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Existing: does this need to be wrapped?

**Comment by chenghao-intel** [comment]
Thank you @cloud-fan for doing this, and I think this can work as a workaround for cases like `Seq[Expression].find`, however, there are still cases like `Map[Expression, Expression]` e.g. in `patterns.scala` as I did in #5798.
Adding new method for `Expression`, probably impact all of `Expression` system (increasing the overhead for maintenance for future?), instead of doing this, I still insist to provide specific collection like `ExpressionMap` or `ExpressionSet`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 4]
```diff
@@ -76,6 +76,19 @@ abstract class Expression extends TreeNode[Expression] {
       case u: UnresolvedAttribute => PrettyAttribute(u.name)
     }.toString
   }
+
+  /**
+   * Returns true if 2 expressions are equal in semantic, which is similar to equals method
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: Returns true when two expressions will always compute the same result, even if they differ cosmetically (i.e. capitalization of names in attributes may be different).

**Comment by marmbrus** [comment]
@chenghao-intel, a single method seems easier to maintain then full Map/Set implementations.  I'll add that its not clear to me the use in patterns is actually invalid.  You can compare expression with equality when you are only trying to find the exact same expression.

I'm merging this to master and 1.4.  I'll improve the wording of the scala doc while merging.

**Comment by chenghao-intel** [comment]
I am not sure why we use the Map[Expression, NamedExpression] in `patterns.scala`, seems `Seq[(Expression, NamedExpression)] is more reasonable, which is definitely confusing for people why we couldn't using `Set/Map` for expressions or should use them is a very ugly way.

**Pull request closed without merging**